File: Top Interview Questions/src/350.intersection-of-two-arrays-ii.py
#
# @lc app=leetcode id=350 lang=python3
#
# [350] Intersection of Two Arrays II
#
from typing import List
# @lc code=start
from collections import defaultdict
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        hashMap = defaultdict(int)
        
        answer = []
        for each in nums1:
            hashMap[each] += 1
        
        for each in nums2:
            if hashMap[each] > 0:
                hashMap[each] -= 1
                answer.append(each)
        return answer
        
# @lc code=end

for list1, list2 in [ [[4,9,5], [9,4,9,8,4]], [[1,2,2,1], [2,2]] ]:
    print(Solution.intersect(None, list1, list2))

File: Top Interview Questions/src/350.intersection-of-two-arrays-ii.java
import java.util.ArrayList;
import java.util.HashMap;

/*
 * @lc app=leetcode id=350 lang=java
 *
 * [350] Intersection of Two Arrays II
 */

// @lc code=start
class Solution {
  public int[] intersect(int[] nums1, int[] nums2) {
    HashMap<Integer, Integer> usedCount = new HashMap<>();
    
    int[] shorter, longer;
    if (nums1.length < nums2.length) {
      shorter = nums1;
      longer = nums2;
    } else {
      shorter = nums2;
      longer = nums1;
    }

    for (int each : shorter) {
      if (usedCount.containsKey(each)) {
        usedCount.put(each, usedCount.get(each) + 1);
      } else {
        usedCount.put(each, 1);
      }
    }
    
    ArrayList<Integer> answerList = new ArrayList<>();
    for (int each : longer) {
      if (usedCount.containsKey(each)) {
        if (usedCount.get(each) > 0) {
          // 카운트 -1
          usedCount.put(each, usedCount.get(each) - 1);
          answerList.add(each);
        }
      } 
    }

    int[] answer = new int[answerList.size()];
    for (int i = 0; i < answer.length; i++) {
      answer[i] = answerList.get(i);
    }
    
    return answer;
  }
}
// @lc code=end

